### PR TITLE
Fix #22240 do not pull all the tags implicitely

### LINKED ISSUE #22240: Implicit pull on run broken on master
Doing `docker run busybox ls` on empty daemon starts pulling all the tags of busybox repo.

Regression point should be #21657 

ping @vdemeester

### PR DESCRIPTION
**Pull request by vdemeester**
This fixes #22240 (`docker run -it busybox /bin/sh` will, on master, pull any tag if `busybox:latest` was not present) <del>in a weird and quick way. I would really prefer the fix in docker/engine-api#205 (in one of the _option_)</del> — it is now cleaner and uses engine-api.

Depends on docker/engine-api#205 :angel: 

/cc @stevvooe @calavera @tonistiigi 

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by vdemeester** [comment]
Updated and build fixed (only `vendor` is red but that's expected before the related engine-api PR is merged).

ping @tonistiigi @stevvooe

File: vendor/src/github.com/docker/engine-api/types/registry/registry.go [hunk 1]
```diff
@@ -82,7 +82,7 @@ type SearchResult struct {
 	IsOfficial bool `json:"is_official"`
 	// Name is the name of the repository
 	Name string `json:"name"`
-	// IsOfficial indicates whether the result is trusted
+	// IsTrusted indicates whether the result is trusted
```
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: This should not be called `IsTrusted`. This field has no bearing on the trust state of an image. Not sure if we need a PR in engine-api to handle this.

Comment (vdemeester) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: Hum, it's already in engine-api, and I guess it's for `golint`, is the attribute is named `IsTrusted` :angel:

Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: This needs to be fixed. These images are not trusted; they are official.

Comment (aaronlehmann) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: This is the comment for the `IsTrusted` flag, not `IsOfficial`. The comment referenced the incorrect variable.

Comment (thaJeztah) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: IIRC IsTrusted was only there for backward compatibility, as it was the old name for official images on Docker Hub; can we officially deprecate and remove it?

Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: @aaronlehmann This was `IsTrusted`, which changed to `IsTrusted` in the code, but had to stay `is_trusted` on the wire. I am not sure about the deprecation status, but let's do so.

Comment (aaronlehmann) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: Wow. Then it definitely is misleading. I thought this referred to content trust. You're right that it needs to be changed.

**Comment by tonistiigi** [comment]
We should add a test for it so we don't break it in the future.

File: vendor/src/github.com/docker/engine-api/types/registry/registry.go (hunk 1 above)
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: Actually, I'm wrong: https://docs.docker.com/engine/reference/api/docker_remote_api_v1.19/.

"This API returns both is_trusted and is_automated images. Currently, they are considered identical. In the future, the is_trusted property will be deprecated and replaced by the is_automated property."

Seems to have been deprecated between 1.19 and 1.20. I propose we remove this from engine-api.

Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: Filed #215

Comment (vdemeester) [inline vendor/src/github.com/docker/engine-api/types/registry/registry.go]: @stevvooe thanks :wink:

**Comment by vdemeester** [comment]
@tonistiigi yes, I'll work on that next few days :angel:

**Comment by vdemeester** [comment]
@tonistiigi from looking at https://github.com/docker/docker/blob/master/integration-cli/docker_cli_pull_test.go#L17 `TestPullFromCentralRegistry` should already cover this case, this sounds strang it didn't..

**Comment by vdemeester** [comment]
@tonistiigi Added a regression test ; it fails on master, and succeed locally.

File: integration-cli/docker_cli_pull_test.go [hunk 2]
```diff
@@ -237,6 +237,21 @@ func (s *DockerHubPullSuite) TestPullAllTagsFromCentralRegistry(c *check.C) {
 	c.Assert(splitLatest, checker.DeepEquals, splitCurrent, check.Commentf("busybox:latest was changed after pulling all tags"))
 }
 
+// TestRunImplicitPullWithNoTagOnlyPullDefaultTag should pull implicitely only the default tag (latest)
+func (s *DockerHubPullSuite) TestRunImplicitPullWithNoTagOnlyPullDefaultTag(c *check.C) {
+	// run with an image we don't have
+	testRequires(c, DaemonIsLinux)
+	out := s.Cmd(c, "run", "busybox")
+	// defer deleteImages("hello-world")
```
Comment (thaJeztah) [inline integration-cli/docker_cli_pull_test.go]: Think this can be removed?

Comment (vdemeester) [inline integration-cli/docker_cli_pull_test.go]: Damn you're right :sweat_smile:

**Comment by thaJeztah** [comment]
LGTM, thanks!

ping @aaronlehmann @tonistiigi PTAL

**Comment by thaJeztah** [comment]
and ping @cpuguy83 as this is blocking your PR :innocent:

File: vendor/src/github.com/docker/engine-api/types/client.go [hunk 3]
```diff
@@ -178,6 +178,7 @@ type ImageLoadResponse struct {
 
 // ImagePullOptions holds information to pull images.
 type ImagePullOptions struct {
+	All           bool
```
Comment (aaronlehmann) [inline vendor/src/github.com/docker/engine-api/types/client.go]: This would benefit from an explanatory comment.

**Comment by aaronlehmann** [comment]
LGTM

File: vendor/src/github.com/docker/engine-api/types/client.go (hunk 3 above)
Comment (vdemeester) [inline vendor/src/github.com/docker/engine-api/types/client.go]: @aaronlehmann true. I'll add a comment on this next PR I do on engine-api :wink:. It should get *in* docker/docker before next release but we shouldn't block this PR for it :angel:

**Comment by cpuguy83** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**